### PR TITLE
Use lsql Lakeview dashboard-as-code implementation to create dashboards

### LINKED ISSUE #1478: [FEATURE]: Use LakeView instead of Redash for dashboards-as-code
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Use LakeView, it's faster

### Proposed Solution

Once LSQL PR is merged, we can start:
- https://github.com/databrickslabs/lsql/pull/66

### Additional Context

_No response_

### LINKED ISSUE #1535: [FEATURE]: Create LakeView counter widget and convert all existing `.sql` counters to it
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Redash is deprecated and takes 45 second per dashboard to deploy. Currently we have 3 dashboards and it doesn't scale to add more of them.

### Proposed Solution

Create LakeView table widget with the ability to make column as a link to notebook or job and migrate existing .sql files to this structure. Implementation has to be based on https://github.com/databrickslabs/lsql/pull/66

### Additional Context

_No response_

### LINKED ISSUE #1536: [FEATURE]: Create LakeView markdown widget and convert all existing `.md` widgets to it
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Redash is deprecated and takes 45 second per dashboard to deploy. Currently we have 3 dashboards and it doesn't scale to add more of them.

### Proposed Solution

Create LakeView table widget with the ability to make column as a link to notebook or job and migrate existing .md files to this structure. Implementation has to be based on https://github.com/databrickslabs/lsql/pull/66

### Additional Context

_No response_

### LINKED ISSUE #1740: [FEATURE] Clickable links in code migration dashboard creation
We want clickable links in the code migration dashboard (#1539), which requires #1478

### PR DESCRIPTION
**Pull request by JCZuurmond**
## Changes
Use lsql Lakeview dashboard-as-code implementation to create dashboards

### Linked issues
Resolves #1478 
Resolves #1535
Resolves #1536
Resolves #1740

### Functionality 

- [x] modified existing workflow: that have dashboards

### Tests

- [x] manually tested
- [x] used existing unit tests: `test_dashboard`
- [x] using exiting integration tests: `test_running_real_migrate_groups_job`
- [x] verified on staging environment (screenshot attached)

### CODE REVIEW

**Comment by JCZuurmond** [comment]
Using the cli first, am able to recreate the main dashboard besides the [filter](https://github.com/databrickslabs/lsql/issues/169)

Screenshot 2024-06-19 at 16 29 15
Screenshot 2024-06-19 at 16 29 24
Screenshot 2024-06-19 at 16 29 34

**Comment by JCZuurmond** [comment]
Main dashboard:
https://github.com/databrickslabs/ucx/assets/5946784/d2b9b3de-de79-4aaa-8cda-4778e48c947d

**Comment by JCZuurmond** [comment]
Works with [lsql@feat/dashboards-as-code-for-ucx](https://github.com/databrickslabs/lsql/tree/feat/dashboard-as-code-for-ucx)
- [ ] Estimates dashboard spacing needs to be improved
- [ ] Clean up unused resources in `dashboards.py`
- [ ] Publish the lakeview dashboard
- [ ] SqlTask.dashboard does not work with Lakeview. The lakeview dashboard "just works", but how do we link it in a specific job

File: src/databricks/labs/ucx/queries/assessment/estimates/03_0_data_migration.md [hunk 1]
```diff
@@ -1,8 +1,7 @@
--- widget title=Table estimates, row=3, col=0, size_x=2, size_y=8
 ## 4 - Data migration to UC
 
-Once you have defined your data model in UC and that you've created appropriate Storage Credentials and External Locations, 
-you can then migrate your data to UC 
+Once you have defined your data model in UC and that you've created appropriate Storage Credentials and External Locations,
+you can then migrate your data to UC
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/queries/assessment/estimates/03_0_data_migration.md:4]: Note that LakeView does not support markdown tables (see below)

File: src/databricks/labs/ucx/queries/assessment/estimates/03_0_data_migration.md [hunk 2]
```diff
@@ -12,17 +16,5 @@ Assumptions for a single table migration estimates:
 - Target Catalogs and schemas has been defined
 - Grants has been defined
 
-Please note that depending on the table type and their location, the migration effort will differ.
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/queries/assessment/estimates/03_0_data_migration.md:15]: Note that LakeView does not support markdown tables, thus converted this to a sql query, see `04_1_data_migration_efforts.sql`

File: src/databricks/labs/ucx/installer/workflows.py [hunk 3]
```diff
@@ -636,14 +636,12 @@ def _job_task(self, task: Task, remote_wheels: list[str]) -> jobs.Task:
...
         assert task.dashboard is not None
         dashboard_id = self._install_state.dashboards[task.dashboard]
-        return replace(
-            jobs_task,
-            job_cluster_key=None,
-            sql_task=jobs.SqlTask(
-                warehouse_id=self._warehouse_id,
-                dashboard=jobs.SqlTaskDashboard(dashboard_id=dashboard_id),
-            ),
+        dashboard_url = f"{self._ws.config.host}/sql/dashboardsv3/{dashboard_id}"
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/installer/workflows.py:639]: Linking a dashboard via a notebook now. SqlTask does not work with Lakeview dashboard

**Comment by JCZuurmond** [comment]
Removed the dashboard link within a job as this is not an option for the Lakeview dashboards (using the SqlTask)

@nfx : The dashboard creation still is not fast when using Lakeview: takes 10-20 seconds.

File: src/databricks/labs/ucx/mixins/fixtures.py [hunk 4]
```diff
@@ -249,6 +249,12 @@ def _path(ws, path):
             [PermissionLevel.CAN_USE, PermissionLevel.CAN_MANAGE],
             _simple,
         ),
+        (
+            "lakeview_dashboard",
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/mixins/fixtures.py:253]: Could someone confirm that this is correctly implemented? I am missing links to docs on what resources should be defined

File: src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql [hunk 5]
```diff
@@ -1,5 +1,17 @@
--- viz type=table, name=Workflow migration problems, search_by=path,code,job_name, columns=path,code,message,workflow_id,workflow_name,task_key,start_line,start_col,end_line,end_col
--- widget title=Workflow migration problems, row=2, col=2, size_x=4, size_y=8
+/*
+--title 'Workflow migration problems'
+--overrides '{"spec": {"encodings": {"columns": [
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql:3]: Added the clickable links by overwriting the `columns` in encodings:


https://github.com/databrickslabs/ucx/assets/5946784/e1ae0954-83f3-4165-88f6-6353935e3aef

Note: it requires to overwrite all columns as the whole list needs to be provided

**Comment by JCZuurmond** [comment]
Clickable links for migrated dashboard `paths` also work

https://github.com/databrickslabs/ucx/assets/5946784/5daec214-e370-4e6f-a8e0-858841b58612

**Comment by JCZuurmond** [comment]
@nfx : Please review the PR. The dashboards look as follows (see video in last comment also):

https://github.com/databrickslabs/ucx/assets/5946784/16f50da0-227d-4ab9-879d-ae03eae79101


https://github.com/databrickslabs/ucx/assets/5946784/f05b90c0-9e08-46c2-849a-d719dd9de244

Screenshot 2024-07-09 at 17 40 13

https://github.com/databrickslabs/ucx/assets/5946784/871bf6a8-d6ae-4270-8464-e88f7b4c49b1

File: src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql [hunk 6]
```diff
@@ -1,5 +1,17 @@
...
+    {"fieldName": "path", "booleanValues": ["false", "true"], "linkUrlTemplate": "/#workspace/{{ @ }}", "linkTextTemplate": "{{ @ }}", "linkTitleTemplate": "{{ @ }}", "linkOpenInNewTab": true, "type": "string", "displayAs": "link", "title": "path"},
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql:4]: Added clickable links for the workspace path using the `#workspace`

**Review by nfx — COMMENTED** [review]
this looks great. Few things to polish before merging

File: src/databricks/labs/ucx/config.py [hunk 7]
```diff
@@ -67,6 +68,16 @@ class WorkspaceConfig:  # pylint: disable=too-many-instance-attributes
     # [INTERNAL ONLY] Whether the assessment should capture only specific object permissions.
     include_object_permissions: list[str] | None = None
 
+    def transform_inventory_database(self, node: sqlglot.Expression) -> sqlglot.Expression:
```
[the remark this answers is not in the document]
Comment (JCZuurmond) [reply src/databricks/labs/ucx/config.py:71]: https://github.com/databrickslabs/lsql/pull/210

File: src/databricks/labs/ucx/installer/workflows.py [hunk 8]
```diff
@@ -492,7 +492,7 @@ def _create_dashboard_links(self, step_name, dashboard_link):
             if len(dashboard_link) == 0:
                 dashboard_link += "Go to the one of the following dashboards after running the job:\n"
             first, second = dash.replace("_", " ").title().split()
-            dashboard_url = f"{self._ws.config.host}/sql/dashboards/{self._install_state.dashboards[dash]}"
+            dashboard_url = f"{self._ws.config.host}/sql/dashboardsv3/{self._install_state.dashboards[dash]}"
```
[the remark this answers is not in the document]
Comment (JCZuurmond) [reply src/databricks/labs/ucx/installer/workflows.py:495]: See https://github.com/databrickslabs/lsql/pull/211

File: src/databricks/labs/ucx/installer/workflows.py (hunk 3 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/installer/workflows.py]: This is outdated, not linking a dashboard in a job anymore as this is not possible with Lakeview

File: src/databricks/labs/ucx/framework/dashboards.py [hunk 9]
```diff
@@ -1,72 +1,15 @@
-import dataclasses
 import logging
```
[the remark this answers is not in the document]
Comment (JCZuurmond) [reply src/databricks/labs/ucx/framework/dashboards.py]: Done, needs the PRs from the other comments to be merged and released

File: src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql [hunk 10]
```diff
@@ -1,5 +1,17 @@
...
+    {"fieldName": "code", "booleanValues": ["false", "true"], "type": "string", "displayAs": "string", "title": "code"},
+    {"fieldName": "message", "booleanValues": ["false", "true"], "type": "string", "displayAs": "string", "title": "message"},
+    {"fieldName": "workflow_id", "booleanValues": ["false", "true"], "linkUrlTemplate": "/jobs/{{ @ }}", "linkTextTemplate": "{{ @ }}", "linkTitleTemplate": "{{ @ }}", "linkOpenInNewTab": true, "type": "string", "displayAs": "link", "title": "workflow_id"},
```
[the remark this answers is not in the document]
Comment (JCZuurmond) [reply src/databricks/labs/ucx/queries/migration/main/02_1_code_compatibility_problems.sql]: done

**Comment by JCZuurmond** [comment]
- [ ] Always deploy the dashboards (in parallel) and remove the `skip_dashboards` flag from the installation context

File: src/databricks/labs/ucx/mixins/fixtures.py (hunk 4 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/mixins/fixtures.py:254]: Done, see integration test and below picture. The permissions were correct, but added a comment with more context

Screenshot 2024-07-11 at 11 02 31

**Comment by JCZuurmond** [comment]
> Always deploy the dashboards (in parallel) and remove the skip_dashboards flag from the installation context

@nfx : Did a quick analysis about how long the `WorkspaceInstallation.run` takes with and without deploying dashboards. After filtering out outliers, i.e. higher/lower than mean plus/minus three times the standard deviation, I get:

| Description | Mean (seconds) | Standard deviation (seconds) | Number of times the test ran |
| -----------  | ---------------- | ----------------------------- | ----------------------------------
| With deploying dashboards | 30.98 | 1.23 | 9 |
| Without deploying dashboards | 32.51 | 1.22 | 8 |

> "without deploying dashboards" ran longer due to the Databricks side, I think, I noticed more outliers when running those tests

It shows that the difference with/without deploying dashboards became negligible. The installation takes a bit more than half a minute

**Comment by JCZuurmond** [comment]
⚠️  I think installation might fail sometimes when the dashboard is created in parallel with the dashboard deployment as lakeview detects that the database does not exists. Need to verify

**Pull request merged**